exporter/awsemfexporter/metric_translator.go

```diff
@@ -90,186 +86,226 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
-		// If metric declarations are defined, extract dimension sets from them
-		dimensions = processMetricDeclarations(metricDeclarations, pmd, labels, rollupDimensionArray)
-	} else {
-		// If no metric declarations defined, create a single dimension set containing
-		// the list of labels
-		dims := labelsSlice
-		if instrumentationLibName != noInstrumentationLibraryName {
-			// If OTel instrumentation lib name is defined, add instrumentation lib
-			// name as a dimension
-			dims = append(dims, oTellibDimensionKey)
-		}
+	// Add on rolled-up dimensions
+	dimensions = append(dimensions, rollupDimensionArray...)
 
-		if len(rollupDimensionArray) > 0 {
-			// Perform de-duplication check for edge case with a single label and single roll-up
-			// is activated
-			if len(labelsSlice) > 1 || (dimensionRollupOption != singleDimensionRollupOnly &&
-				dimensionRollupOption != zeroAndSingleDimensionRollup) {
-				dimensions = [][]string{dims}
-			}
-			dimensions = append(dimensions, rollupDimensionArray...)
-		} else {
-			dimensions = [][]string{dims}
+	metrics := make([]map[string]string, len(groupedMetric.Metrics))
+	idx = 0
+	for metricName, metricInfo := range groupedMetric.Metrics {
+		metrics[idx] = map[string]string{
+			"Name": metricName,
+			"Unit": metricInfo.Unit,
 		}
+		idx++
 	}
 
-	// Build list of CW Measurements
-	var cwMeasurements []CwMeasurement
-	if len(dimensions) > 0 {
-		cwMeasurements = []CwMeasurement{
-			{
-				Namespace:  namespace,
-				Dimensions: dimensions,
-				Metrics:    metricSlice,
-			},
+	return CWMeasurement{
+		Namespace:  groupedMetric.Metadata.Namespace,
+		Dimensions: dimensions,
+		Metrics:    metrics,
+	}
+}
+
+// groupedMetricToCWMeasurementsWithFilters filters the grouped metric using the given list of metric
+// declarations and returns the corresponding list of CW Measurements.
+func groupedMetricToCWMeasurementsWithFilters(groupedMetric *GroupedMetric, config *Config) (cWMeasurements []CWMeasurement) {
+	labels := groupedMetric.Labels
+
+	// Filter metric declarations by labels
+	metricDeclarations := make([]*MetricDeclaration, 0, len(config.MetricDeclarations))
+	for _, metricDeclaration := range config.MetricDeclarations {
+		if metricDeclaration.MatchesLabels(labels) {
+			metricDeclarations = append(metricDeclarations, metricDeclaration)
 		}
 	}
 
-	timestampMs := time.Now().UnixNano() / int64(time.Millisecond)
-	if dp.TimestampMs > 0 {
-		timestampMs = dp.TimestampMs
+	// If the whole batch of metrics don't match any metric declarations, drop them
+	if len(metricDeclarations) == 0 {
+		labelsStr, _ := json.Marshal(labels)
+		metricNames := make([]string, 0)
+		for metricName := range groupedMetric.Metrics {
+			metricNames = append(metricNames, metricName)
+		}
+		config.logger.Debug(
+			"Dropped batch of metrics: no metric declaration matched labels",
+			zap.String("Labels", string(labelsStr)),
+			zap.Strings("Metric Names", metricNames),
+		)
+		return
```

Comment:
I am thinking what should be the expected user experience here.
1. Based on README, `A label_matcher section defines a matching rule against the labels of the incoming metric. Only metrics that match the rules will be used by the surrounding metric_declaration.` It seems it determines if the metric_declaration should apply on this metric, instead of dropping the entire metric. So should we drop it or skip it here?
2. If we treat it as a filter which drops the metric not matching, the expected output metrics should only contain the dimensions declared inside label_matcher. If I understand correctly, Currently the `rollupDimensionArray (line254)` is calculated based on original `labels` which include not only extracted dimensions. Is this experience expected?

Reply:
> Based on README, A label_matcher section defines a matching rule against the labels of the incoming metric. Only metrics that match the rules will be used by the surrounding metric_declaration. It seems it determines if the metric_declaration should apply on this metric, instead of dropping the entire metric. So should we drop it or skip it here?

The `label_matcher` use cases were mainly for the workload metrics(like NGINX workload Prometheus metrics) which produces lots of metrics. Customers can use `label_matcher` to generate the metrics which has certain labels. All the un-matched metrics won't generate `CloudWatchMetrics` section in EMF logs but just keep all the dimensions and Metric name and value as plain logs entries when sending to CloudWatch backend.

Reply:
>If we treat it as a filter which drops the metric not matching, the expected output metrics should only contain the dimensions declared inside label_matcher. If I understand correctly, Currently the rollupDimensionArray (line254) is calculated based on original labels which include not only extracted dimensions. Is this experience expected?

I think we inherit this behavior from CloudWatch Agent Prometheus support [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/ContainerInsights-Prometheus-Setup-configure.html). `label_matcher` was used to help customer to determine what metrics they want but not truncate any dimensions when send it out. I think we need to turn off the default dimension Rollup in EMF exporter. Let's discuss it with Product team.